nipoppy/zenodo_api.py

```diff
@@ -0,0 +1,250 @@
+        )
+        self.access_token = access_token
+
+        if logger is None:
+            self.logger = get_logger(__name__)
+        else:
+            self.logger = logger
+
+        # Access token is required for uploading files
+        if self.access_token is not None:
+            self.headers = {
+                "Authorization": f"Bearer {self.access_token}",
+            }
+        else:
+            self.headers = {}
+
+    def set_logger(self, logger: logging.Logger):
+        """Set the logger for the ZenodoAPI instance."""
+        self.logger = logger
+
+    def download_record_files(self, record_id: str, output_dir: Path):
+        """Download the files of a Zenodo record in the `output_dir` directory.
+
+        Parameters
+        ----------
+        record_id : str
+            Record ID in Zenodo.
+        output_dir : Path
+            Output directory to save the files.
+
+        Raises
+        ------
+        InvalidChecksumError
+            Checksum mismatch between the downloaded file and the expected checksum.
+        """
+        record_id = record_id.removeprefix("zenodo.")
+        output_dir.mkdir(parents=True, exist_ok=True)
+
+        response = httpx.get(
+            f"{self.api_endpoint}/records/{record_id}/files",
+            headers=self.headers,
+        )
+        if response.status_code != 200:
+            raise ZenodoAPIError(
+                f"Failed to get files for zenodo.{record_id}: {response.json()}"
+            )
+
+        # Exclude "md5:" prefix
+        files = {
+            entry["key"]: entry["checksum"].removeprefix("md5:")
+            for entry in response.json()["entries"]
+        }
+        for file, checksum in files.items():
+            response = httpx.get(
+                f"{self.api_endpoint}/records/{record_id}/files/{file}/content",  # noqa E501
+                headers=self.headers,
+            )
+            if response.status_code != 200:
+                raise ZenodoAPIError(
+                    f"Failed to download file for zenodo.{record_id}: {file}"
+                    f"\n{response.json()}"
+                )
+
+            # Checksum verification before writing to disk
+            content_md5 = hashlib.md5(response.content).hexdigest()
+            if content_md5 != checksum:
+                raise InvalidChecksumError(
+                    "Checksum mismatch: " f"'{file}' has invalid checksum {content_md5}"
+                )
+
+            output_dir.joinpath(file).write_bytes(response.content)
+
+    def _create_new_version(self, record_id: str, metadata: dict) -> str:
+        response = httpx.post(
+            f"{self.api_endpoint}/records/{record_id}/versions",
+            headers=self.headers,
+        )
+        if response.status_code != 201:
+            raise ZenodoAPIError(
+                f"Failed to create a new version for zenodo.{record_id}:"
+                + response.json()
+            )
```

Comment:
Invalid use of `+` operator, this is what we get when test coverage is low 😅 
```suggestion
            raise ZenodoAPIError(
                f"Failed to create a new version for zenodo.{record_id}:"
                f" {response.json()}"
            )
```